3단계 - 배포 스크립트 작성하기

tlaqk229 opened the pull request:
안녕하세요 😄 
3단계 미션 완료하여 PR드립니다!
이번 미션은 저에게 참 힘들었던 미션이라 감회가 깊네요 😭 
네트워크랑 인프라에 대한 지식이 워낙 부족하다 보니 매 단계마다 찾아보고 공부할게 참 많기도 했고
뭔가 이상하게 할 땐 알아볼 것도 많고 신경쓸 것도 많은데 다하고 나면 결과물이 너무 없는 것 같아 허무하기도 했구요.. 😥 

그래도 이번 기회에 몰랐던 부분 다 짚고 넘어가자 싶어 오기로 미션구현에 치중하기 보다는 모르는 내용을 찾아보고
실습해보는 데 힘써보았는데, 오랜만에 새로운 것들을 익히고 이것저것 시도해보고 하다보니 재밌는 시간 보냈습니다 💫 
리눅스를 거의 사용해본적이 없어서 1단계, 2단계도 쉽지 않았고 명령어나 사용방법이나 여러가지로 찾아볼게 많았지만
3단계는 스크립트 작성이다보니 거의 bash와 붙어살았던 것 같네요ㅋㅋ 조금 친해진 것 같아 뿌듯하고
찾아보고 연습해본 것 토대로 스크립트 나름대로 신경 많이 써서 작성해봤는데 부족한 점이 많은 것 같습니다ㅜㅜ
이상한 부분, 모자란 부분 가감없이 코멘트 부탁드려요! 🙏 

스크립트 파일은 /script 아래 deploy.sh라는 이름으로 넣어두었습니다!
요구사항에 반복적으로 실행하더라도 정상 배포하는 스크립트를 작성해보라는 내용이 있어서 여러번 실행하면서
예상하지 못한 부분에서 문제가 생기는 부분들을 처리해보려고 노력했는데요
가령 check_diff에서 빠른 테스트를 위해 pull, build를 주석처리해두고 테스트해보니
프로세스가 채 죽기도 전에 구동 시작명령이 들어가서 서비스가 시작되지 않는 문제가 있더라구요
그래서 거기에 프로세스가 종료되었는지 확인하는 코드를 넣는다든지 하는.. 그러다보니 코드가 복잡하고 길어진것같아요.
그런데 스크립트는 기존 자바라든지 프로그래밍 언어와는 달리 함수 분리해서 결과값 리턴하는 것이 맘대로 잘안되더라구요.
그래서 나름대로 함수 분리를 한다고 해보았는데 여전히 복잡해보이네요 🌀 
실제로는 배포 스크립트파일 어떤식으로 작성되는지 궁금합니다! 예외적인 부분들을 잡는 코드가 많이 들어가는지,
아니면 핵심적인 명령만 작성하고 모니터링에 조금 더 집중하게되는지.. 중점적으로 봐야할 부분 가이드 주시면 감사하겠습니다!

마지막으로 shell script작성법과 cron, crontab에 대해 공부한 내용 나중에 다시 복기하기 좋도록 정리해봤는데요.
혹시 잘못되었거나, 이런 내용도 덧붙여주면 좋을 것 같다싶은 부분 있다면 코멘트 부탁드립니다! 🙇 

항상 빠른 리뷰 감사드려요! 마지막 단계 리뷰도 잘 부탁드립니다~! 🚀

tlaqk229 committed:
docs(study): shell script, cron&crontab 학습내용 정리

tlaqk229 committed:
docs(script): deploy script 작성

tlaqk229 committed:
docs(script): 스크립트 수정

- 스크립트에 첫번째 인자로 check_diff 넘겨주는 경우 모드 추가입력가능
- 오류 등 중간에 종료되는 경우에도 스크립트 종료문구 출력되도록 수정
- 잘못 명명된 함수명 수정

tlaqk229 committed:
docs(mission): 3단계 미션수행 질문 답변 및 요구사항 작성

seokhongkim commented on docs/cron&crontab_정리.md:
```diff
@@ -0,0 +1,158 @@
+# Cron & Crontab
```
cron, redirect 에 대해서 많이 학습하고 잘 정리하셨네요!

redirect 의 경우 출력하지 않고 싶은 내용을 버리기 위해 `/dev/null` (bit bucket) 로 redirect 하기도 하니, 이 내용도 같이 학습해보셔도 좋겠습니다. :)

tlaqk229 replied:
bit bucket 이란 멋진 용어가 있는지 몰랐네요. 감사합니다 😁  내용 찾아보고 추가해봤습니다!

seokhongkim commented on docs/shell_script_정리.md:
```diff
@@ -0,0 +1,231 @@
+# Shell Script
```
쉘 스크립트 작성을 위해 학습하신 내용도 잘 정리해주셨습니다. 👍 
이 페이지만 참고해도 쉘 스크립트는 바로 작성을 시작할 수 있겠네요!

seokhongkim commented on docs/step3.md:
````diff
@@ -0,0 +1,76 @@
+```
+* * * * * /home/ubuntu/deploy.sh check_diff prod >> /home/ubuntu/nextstep/logs/deploy_sh.log 2>&1
+```
+- 매분마다 git 원격저장소 차이 비교 후 차이 있는 경우 반영, 재시작
````
스크립트 작성 및 cron 작업을 통한 빌드 자동화 설정도 잘 해주셨습니다. 👍 
서비스가 커지면 build 시간이 길어질 수 있는데요. 이 경우 매 분마다 cron 이 실행되면 여러 개의 빌드가 동시에 진행되는 순간이 발생할 수 있고, 빌드 과정에 의도치 않은 문제가 생길 수 있을 것 같습니다.
약간의 버퍼 간격을 두고 cron 작업이 진행되어도 좋을 것 같습니다. :)

tlaqk229 replied:
그 부분은 미처 생각지 못했네요!❗ 5분마다 실행하는 것으로 변경해봤습니다 😃

seokhongkim commented on script/deploy.sh:
```diff
@@ -0,0 +1,284 @@
+  echo -e "${txtcyn}check if there is a difference between remote repo and local repo${txtrst}"
+  cd "$APP_PATH$BUILD_FOLDER_NAME" || exit_script
+  git fetch origin
+  master=$(git rev-parse $CURRENT_BRANCH)
```
로컬 브랜치 hash 값이니 변수명을 local 로 변경해보면 어떨까요? :)

tlaqk229 replied:
무의식적 습관이 이렇게 무섭군요.. 😓  수정했습니다!

seokhongkim commented on script/deploy.sh:
```diff
@@ -0,0 +1,284 @@
+#!/bin/bash
```
기능 단위로 함수를 잘 만들어 주셨고, 중간에 예외가 발생할 수 있는 케이스도 꼼꼼하게 처리해주셨네요.
진행중인 작업에 대한 로그 메시지도 잘 작성해주셨습니다. 👍

seokhongkim commented on docs/step3.md:
````diff
@@ -0,0 +1,76 @@
+   - 사용자에게 스크립트 실행 시 전달 가능한 인자(기능목록) 안내
+
+```
+※ app 디렉토리와 build 디렉토리 이중 운영 이유
````
코드를 가져오기 전 기존 버전의 코드를 백업하시는 과정 좋습니다. 👍 
문제가 발생했을 때 백업된 코드만 빌드해서 서비스를 띄우면 빠른 롤백이 가능하겠네요.

실행중인 서비스 코드에 대해서는 따로 관리가 필요하지 않을 것 같기도 하네요. :)
실행할 jar 파일만 복사해서 app 경로에서 관리해도 괜찮을 것 같습니다.

tlaqk229 replied:
생각해보니 그러네요! 백업본까지 하면 괜히 너무 많은 복사본을 갖고 있게되네요 😔
app 디렉토리에는 jar만 보관하도록 하고 변경소스 반영 시 build 디렉토리에서 jar파일만 덮어씌우도록 수정해봤습니다!

그리고 기존에는 백업 -> 반영 -> 백업본 삭제 하도록 구성해두었었는데 코멘트 읽고 보니
반영은 잘되었지만 기능에 문제가 있어서 이전버전으로 롤백해야할 때 정작 백업본이 없는 상황이라
기존 백업본 삭제(있는경우) -> 백업 -> 반영 순으로 수정해봤습니다 ✏️

seokhongkim commented:
먼저 질문 남겨주신 부분에 대해 제 생각 말씀 드립니다. :)

> 실제로는 배포 스크립트파일 어떤식으로 작성되는지 궁금합니다! 예외적인 부분들을 잡는 코드가 많이 들어가는지,
아니면 핵심적인 명령만 작성하고 모니터링에 조금 더 집중하게되는지.. 중점적으로 봐야할 부분 가이드 주시면 감사하겠습니다!

제 경험상으로는 회사의 상황, 시스템 성숙도, 인프라, 조직 구조에 따라 매번 달랐습니다.
스크립트를 사용하는 경우, 현재 작성해주신 스크립트보다 간단한 스크립트를 작성하고, 서버에 들어가서 배포를 진행하는 경우도 있었습니다. 배포해야 할 서버가 늘어남에 따라, 서버에 접속하지 않고 외부(배포 서버)에서 특정 서버에 배포하도록 스크립트를 고도화 하기도 합니다. 두 과정 모두 운영 배포의 경우 무중단 배포가 필요하기에, 현재 배포 로직에서 약간의 변경(기존 서비스를 유지한 상태에서 새로운 서비스를 띄우고, 포트 포워딩 또는 nginx 설정을 변경한 후, 기존 서비스를 종료)이 필요합니다.
스크립트를 사용하지 않고 클라우드 업체에서 제공하는 툴을 사용해서 배포를 하기도 합니다. aws 를 예로 들면 beanstalk 을 예로 들 수 있겠네요. 무중단 배포도 ELB를 이용하여 진행할 수 있습니다.
이외에 docker 를 사용하는 경우 이미지를 빌드하는 것과 서비스를 띄우는 과정이 분리되기도 하겠네요.
큰 틀에서 봤을 때는 배포 과정에서 고려하는 내용, 배포 진행 단계는 비슷했던 것 같고, 구현에서 보면 많이 달라지는 것 같네요. :)

회사에 모니터링과 배포 도구 관련된 부분을 전담하는 팀이 있으면, 두 팀이 각각의 역할을 별도로 진행하는 것을 봤습니다. 개별 팀에서 배포, 모니터링도 각자 처리하면 우선 순위에 따라 진행했는데요. 이 경우에는 모니터링을 더 중요하게 고려했습니다. 배포는 최소한의 스크립트만 작성해서 사용했습니다.

seokhongkim reviewed (changes requested):
윤지님, 3단계 미션 잘 진행해주셨습니다. 👍 
인프라 미션은 학습하고 설정하는 내용은 많은데 비해, 결과물이 확 눈에 띄지는 않는 것 같아요.
이전 단계 미션 진행 후 허무하기도 하셨다고 했는데, 충분히 이해되네요.

cron, shell script 작성 관련하여 학습하신 내용을 잘 정리해주셨네요!
이 내용만 학습해도 도움이 많이 될 것 같습니다.

질문 남겨주신 내용과 스크립트에 대해 코멘트를 조금 남겼습니다.
코멘트 확인 부탁드립니다. 🙏

tlaqk229 committed:
docs(study): bit bucket('/dev/null')에 대한 내용 추가

tlaqk229 committed:
docs(script): 의미에 맞게 변수명 수정 (현재브랜치hash값 담는 변수)

tlaqk229 committed:
docs(script): 가독성 향상 및 메세지 성격에 따른 텍스트 속성 부여

- 처리 : Cyan
- 상태, 속성 안내 : Yellow
- 오류 : Red
- 결과 안내 및 그에 따른 다음 처리 : Magenta Background
- 처리 실패 안내 : Light Red Background

tlaqk229 committed:
docs(script): 변경된 소스 반영 로직 수정

- 기존: 백업 -> 반영 -> 백업본 삭제
- 수정: 기존 백업본 삭제 -> 현재 소스 백업 -> 반영

- 반영 완료 후 백업본을 삭제하지 않고 다음 반영 전까지 유지함으로써
  반영한 버전에 문제가 있을 경우 백업본을 사용, 빠른 롤백이 가능하도록함

tlaqk229 committed:
docs(script): 서비스 구동을 위한 디렉토리에 jar만 관리하는 방식으로 수정

tlaqk229 committed:
docs(script): 소스 재반영 시간 기록 추가, 서비스 구동 로그 기록관련 수정

- 원격 레포지토리와 비교하여 차이점 발견하여 반영시작하는 시간 출력하도록 수정
- 서비스 구동 로그가 구동 시 마다 새로 작성되게 되어있어 추가기록하는 방식으로 수정

tlaqk229 committed:
docs(mission): 스크립트 및 cron설정 변경사항 반영

seokhongkim reviewed (approved):
윤지님, 3단계 미션 코멘트도 잘 반영해주셨습니다. 👍 
3단계 PR도 머지하겠습니다.

남은 교육 기간 동안에도 지금처럼 열심히 학습하고 정리하시면, 목표로 정하신 성장과 성취 모두 이루실 수 있을 것 같아요. 응원하겠습니다. 💪 
인프라 미션 수고 많으셨습니다!
